ecosystem/cache/handlers.go

```diff
@@ -64,17 +66,42 @@ func (cv *LastestCacheStore) Cost() int64 {
 	return 8 + 16
 }
 
+func (s *RelayerCacheServer) getSeenBlockForSharedStateMode(chainId string, sharedStateId string) int64 {
+	if sharedStateId != "" {
+		id := latestBlockKey(chainId, sharedStateId)
+		value, found := getNonExpiredFromCache(s.CacheServer.finalizedCache, id)
+		if !found {
+			utils.LavaFormatInfo("Failed fetching state from cache for this user id", utils.LogAttr("id", id))
+			return 0 // we cant set the seen block in this case it will be returned 0 and wont be used in the consumer side.
+		}
+		utils.LavaFormatInfo("getting seen block cache", utils.LogAttr("id", id), utils.LogAttr("value", value))
+		if cacheValue, ok := value.(int64); ok {
+			return cacheValue
+		}
+		utils.LavaFormatError("Failed converting cache result to int64", nil, utils.LogAttr("value", value))
+	}
+	return 0
+}
+
 func (s *RelayerCacheServer) GetRelay(ctx context.Context, relayCacheGet *pairingtypes.RelayCacheGet) (cacheReply *pairingtypes.CacheRelayReply, err error) {
 	requestedBlock := relayCacheGet.Request.RequestBlock // save requested block
-
 	cacheReply, err = s.getRelayInner(ctx, relayCacheGet)
+
 	var hit bool
 	if err != nil {
 		s.cacheMiss(ctx, err)
+		// cacheReply is nil as we didn't get a hit. we still need to set the seen block.
+		if cacheReply == nil {
+			cacheReply = &pairingtypes.CacheRelayReply{}
+		}
 	} else {
 		hit = true
 		s.cacheHit(ctx)
 	}
+
+	// set seen block if required
+	cacheReply.SeenBlock = s.getSeenBlockForSharedStateMode(relayCacheGet.ChainID, relayCacheGet.SharedStateId)
```

Comment:
we are doing the two cache reads sequentially doubling the delay, lets do a wait group and do both at the same time